### ISSUE TITLE
SM 3.2 - SCTOOL progress percentage over 100% 

### ISSUE THREAD

**Issue by ppalczak** [scylladb/scylla-manager#3534]
##Problem
After upgrade to SM 3.2 Sctool displays progress over 100%

Status:		RUNNING
Start time:	23 Aug 23 08:12:53 UTC
Duration:	2h37m24s
Progress:	103%

```

├───────────────────────────────┼─────────────────────────────────────┼──────────┼────────────┤
│ system_auth                   │                     role_attributes │ 123%     │ 214h42m23s │
│ system_auth                   │                        role_members │ 123%     │ 214h42m53s │
│ system_auth                   │                    role_permissions │ 123%     │ 214h43m59s │
│ system_auth                   │                               roles │ 123%     │ 214h43m25s │
├───────────────────────────────┼─────────────────────────────────────┼──────────┼────────────┤
│ system_distributed_everywhere │      cdc_generation_descriptions_v2 │ 307200%  │ 7m26s      │
├───────────────────────────────┼─────────────────────────────────────┼──────────┼────────────┤
│ system_distributed            │         cdc_generation_descriptions │ 123%     │ 214h46m7s  │
│ system_distributed            │           cdc_generation_timestamps │ 123%     │ 214h45m3s  │
│ system_distributed            │         cdc_streams_descriptions_v2 │ 123%     │ 214h46m38s │
│ system_distributed            │                      service_levels │ 123%     │ 214h44m30s │
│ system_distributed            │                   view_build_status │ 123%     │ 214h45m34s │
╰───────────────────────────────┴─────────────────────────────────────┴──────────┴────────────╯
```
One table reached:
| system_distributed_everywhere |      cdc_generation_descriptions_v2 | 307200%  | 2h32m27s   |

**Comment by Michal-Leszczynski** [comment]
Was this repair paused and resumed?

**Comment by ppalczak** [comment]
> 

no, it wasn't paused and resumed. 
It's still in progress :
Run:		dc1ea237-418c-11ee-a456-4201ac1e0012
Status:		RUNNING
Start time:	23 Aug 23 08:12:53 UTC
Duration:	3h52m57s
Progress:	108%

@vladzcloudius adding you to the loop

**Comment by Michal-Leszczynski** [comment]
This scenario can be replicated like that:
- run repair task on SM 3.1
- wait for it to finish successfully 
- upgrade SM to 3.2
- start the same repair task

SM 3.2 introduced some schema changes in manager DB and it looks like it is causing this problem.
What is happening, is that SM thinks that the previous run of this repair task hasn't finished and it resumes it (you can also see it by the long duration time).

SM has 2 tables for tracking repair progress:
- `repair_run_state` - used for resuming paused repair, keeps track of finished ranges
- `repair_run_progress` - used for aggregating and printing progress

Changes applied to the first table don't create any problem - this repair task will basically run from scratch which is what we want in this scenario.

Changes applied to the second table don't have any impact on the repair task procedure, but it affects the `sctool progress` command output.

So basically, only the first run of repair task created before SM 3.2 should be affected and only in terms of printing progress.

**Comment by ppalczak** [comment]
> This scenario can be replicated like that:
> 
> * run repair task on SM 3.1
> * wait for it to finish successfully
> * upgrade SM to 3.2
> * start the same repair task
> 
> SM 3.2 introduced some schema changes in manager DB and it looks like it is causing this problem. What is happening, is that SM thinks that the previous run of this repair task hasn't finished and it resumes it (you can also see it by the long duration time).
> 
> SM has 2 tables for tracking repair progress:
> 
> * `repair_run_state` - used for resuming paused repair, keeps track of finished ranges
> * `repair_run_progress` - used for aggregating and printing progress
> 
> Changes applied to the first table don't create any problem - this repair task will basically run from scratch which is what we want in this scenario.
> 
> Changes applied to the second table don't have any impact on the repair task procedure, but it affects the `sctool progress` command output.
> 
> So basically, only the first run of repair task created before SM 3.2 should be affected and only in terms of printing progress.

Thank you @Michal-Leszczynski for a clear explanation. We'll inform customer about that

**Comment by doctorg-ml** [comment]
I have created ticket [12810](https://cloudsupportscylla.zendesk.com/agent/tickets/12810) for Sleeper to link this issue.

**Comment by vladzcloudius** [comment]
> This scenario can be replicated like that:
> 
>     * run repair task on SM 3.1
> 
>     * wait for it to finish successfully
> 
>     * upgrade SM to 3.2
> 
>     * start the same repair task
> 
> 
> SM 3.2 introduced some schema changes in manager DB and it looks like it is causing this problem. What is happening, is that SM thinks that the previous run of this repair task hasn't finished and it resumes it (you can also see it by the long duration time).
> 
> SM has 2 tables for tracking repair progress:
> 
>     * `repair_run_state` - used for resuming paused repair, keeps track of finished ranges
> 
>     * `repair_run_progress` - used for aggregating and printing progress
> 
> 
> Changes applied to the first table don't create any problem - this repair task will basically run from scratch which is what we want in this scenario.
> 
> Changes applied to the second table don't have any impact on the repair task procedure, but it affects the `sctool progress` command output.
> 
> So basically, only the first run of repair task created before SM 3.2 should be affected and only in terms of printing progress.

Michal, could you, please, estimate when we can see the fix for this if at all (see more below)?
Being able to see an adequate progress is an important SM functionality.

In general it's expected that the upgrade would not cause what is described here.

If dropping and creating repair (and maybe other?) task(s) anew would work this issue around it would make this issue minor but still an issue.

Could you, please, confirm if that would be a working workaround?

**Comment by Michal-Leszczynski** [comment]
> In general it's expected that the upgrade would not cause what is described here.

That's true, I guess it was missed in upgrade tests.

> However if dropping and creating repair (and maybe other?) task(s) anew would work this issue around it would make this issue minor but still an issue.
Could you, please, confirm if that would be a working workaround?

Recreating repair tasks should solve this issue.
But it's still an issue and I will create a fix that does not require this additional action.

**Comment by Michal-Leszczynski** [comment]
Or just running the first repair task after upgrade with `--continue=false` should also be a good work around.

**Comment by vladzcloudius** [comment]
> Or just running the first repair task after upgrade with `--continue=false` should also be a good work around.

If it's only relevant to tasks that were paused and then resumed after the upgrade then this reduces the "badness" of this issue even further.

AFAIU this in particular means that  the impacted repair task would only hit this issue once after the upgrade and next time it runs from the beginning it will no longer show symptoms mentioned above, right?

@ppalczak FYI.

**Comment by tzach** [comment]
> If it's only relevant to tasks that were paused and then resumed after the upgrade then this reduces the "badness" of this issue even further.

If this is the case, we can add a known issue to the release notes, and/or upgrade doc

**Comment by Michal-Leszczynski** [comment]
> If it's only relevant to tasks that were paused and then resumed after the upgrade then this reduces the "badness" of this issue even further.

Unfortunately, that's not the case. It can also be replicated like that:
> This scenario can be replicated like that:
>- run repair task on SM 3.1
>-  wait for it to finish successfully
>- upgrade SM to 3.2
>- start the same repair task

So it's not only about paused tasks.

> AFAIU this in particular means that the impacted repair task would only hit this issue once after the upgrade and next time it runs from the beginning it will no longer show symptoms mentioned above, right?

That's correct.

**Comment by vladzcloudius** [comment]
> > If it's only relevant to tasks that were paused and then resumed after the upgrade then this reduces the "badness" of this issue even further.
> 
> Unfortunately, that's not the case. It can also be replicated like that:
> 
> > This scenario can be replicated like that:
> > 
> > * run repair task on SM 3.1
> > * wait for it to finish successfully
> > * upgrade SM to 3.2
> > * start the same repair task
> 
> So it's not only about paused tasks.
> 
> > AFAIU this in particular means that the impacted repair task would only hit this issue once after the upgrade and next time it runs from the beginning it will no longer show symptoms mentioned above, right?
> 
> That's correct - the underlying reason for this issue is that SM 3.2 thinks that previous task run from SM 3.1 hasn't finished and it tries to resume it (even when the previous run has finished with success). SM ends up repairing all token ranges from the scratch, but it copies progress of previous repair task and adds it to the current repair progress and that's why we see over the 100% progress. That's also the reason why running this repair task once with `sctool start --no-continue` should also be a good workaround.

I see. Thanks for a detailed explanation.
@tzach, yeah, IMO this issue can definitely be put in the "known issues" section in the release notes. It's super-minor IMO.
